llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -5838,9 +5840,21 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
     return SDValue();
   }
   case Intrinsic::experimental_cttz_elts: {
-    SDValue NewCttzElts =
-        DAG.getNode(AArch64ISD::CTTZ_ELTS, dl, MVT::i64, Op.getOperand(1));
+    SDValue CttzOp = Op.getOperand(1);
+    EVT VT = CttzOp.getValueType();
+
+    if (!VT.isScalableVector()) {
+      // Retrieve original fixed-width vector from ISD::TRUNCATE Node.
+      assert(CttzOp.getOpcode() == ISD::TRUNCATE && "Expected ISD::TRUNCATE!");
```

Comment:
Thanks for pointing this out - I have updated so that we sign extend to the correct type before passing this (now valid) mask to `convertFixedMaskToScalableVector`.

We don't actually manage to optimise away the sign_extend, at present - I have updated the regression tests to reflect this as it seems to make sense to fix this in a separate, follow up patch?

Reply:
Agreed, the DAG combine should be a separate PR.